ARROW-11677: [C++][Docs] Add basic C++ datasets documentation

lidavidm opened the pull request:
This is mostly based on the Python documentation. This also adds a new C++ example to accompany the documentation.

lidavidm commented:
Rendered (though Github is acting up):
- https://lidavidm.github.io/arrow-docs-next/cpp/dataset.html#tabular-datasets
- https://lidavidm.github.io/arrow-docs-next/cpp/api/dataset.html#dataset

pitrou commented on docs/source/cpp/api/dataset.rst:
```diff
@@ -0,0 +1,165 @@
+Interface
+=========
+
+.. doxygenclass:: arrow::dataset::Fragment
```
Note that you may use `doxygengroup`s to reduce the burden of listing and maintaining individual APIs here.

lidavidm commented:
The example is a bit long because it does the same thing several times with minor variations, with most things duplicated each time so snippets can be used as standalone examples in the docs. However the example could certainly be better structured/documented so it's readable on its own as well.

drin commented:
@lidavidm I'm curious on your thoughts about other things that `ScannerBuilder`s can be constructed from and whether this is something we would want to add to these commits, or if we should have a separate commit?

As an example, one of the things that was difficult to figure out for our research group from existing documentation was how to apply a scan to something already in memory:

```cpp
    using InMemDatasetPtr = std::shared_ptr<arrow::dataset::InMemoryDataset>;
    using ScanBuilderPtr  = std::shared_ptr<arrow::dataset::ScannerBuilder>;

    ...

    // Create an in-memory dataset from the parsed record batch
    InMemDatasetPtr batch_as_dataset = std::make_shared<arrow::dataset::InMemoryDataset>(
         table_batch->schema()
        ,RecordBatchVec({ table_batch })
    );

    // Create a scanner to pass the expression
    ARROW_ASSIGN_OR_RAISE(
         ScanBuilderPtr scanbuilder
        ,batch_as_dataset->NewScan()
    );

    ...
```

lidavidm commented:
@drin thanks for the feedback. I can add a snippet about that as well. If there are other things that were confusing or difficult to figure out, I'd like to tackle them too.

drin commented on cpp/examples/arrow/dataset-documentation-example.cc:
```diff
@@ -217,24 +229,29 @@ std::shared_ptr<arrow::Table> SelectAndProjectDataset(
+  // Also derive a new column.
   names.push_back("b_large");
   exprs.push_back(ds::greater(ds::field_ref("b"), ds::literal(1)));
   ABORT_ON_FAILURE(scan_builder->Project(exprs, names));
```
From what I've seen, we've tried to do this directly on the ScannerBuilder:

```cpp
    std::vector<std::string>   projection_attrs;
    arrow::dataset::Expression selection_expr;

    ...

    ARROW_RETURN_NOT_OK(scanbuilder->Project(projection_attrs));
    ARROW_RETURN_NOT_OK(scanbuilder->Filter(selection_expr));

    ...
```

I am looking at the code first, so I may just need to read the actual documentation included in this PR, but it may be nice to include a comment that references the documentation of why to prefer one approach to another or if they're equivalent mechanisms.

lidavidm replied:
Ah, the example here is using Project to define new virtual columns from the physical columns, while your snippet is simply selecting a subset of the physical columns to read. The former API is a superset of the latter. There's an example of what you currently do a little bit prior to this in the file. I've added explanatory comments for these examples, though in the prose I think it should also be clear.

drin commented:
> @drin thanks for the feedback. I can add a snippet about that as well. If there are other things that were confusing or difficult to figure out, I'd like to tackle them too.

Sounds great! I appreciate your work. I have only recently gathered enough experience to be able to express what is/was difficult and why, so I'll try to keep adding my thoughts here (and directly in JIRA).

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 151-170
```
Maybe stop at line 165 in this snippet?

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+   :linenos:
+   :lineno-match:
+
+Here, we're also passing the filesystem to use and the file format to
```
`Here, we're` -> `We're`

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+   :lineno-match:
+
+Here, we're also passing the filesystem to use and the file format to
+read. This lets us choose between things like reading local files or files in
```
`and the file format to read` -> `and the file format to use for reading`

westonpace replied:
`choose between things like reading` -> `choose between reading`

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+read. This lets us choose between things like reading local files or files in
+Amazon S3, or between Parquet and CSV.
+
+In addition to a base directory path, we can list file paths manually.
```
`a base directory path` -> `searching a base directory`

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+In addition to a base directory path, we can list file paths manually.
+
+Creating a :class:`arrow::dataset::Dataset` in this way loads nothing into
+memory; it only crawls the directory to find all the files
```
"loads nothing into memory" seems a little misleading.  Wouldn't it be more accurate to say "does not begin reading the data"?  Also what is `in this way` referring to?  Isn't this statement true whether you are creating a dataset from a directory or from a list of paths?  In fact, you might want to explain why it is crawling the directory to find all the files given we just presumably listed the paths manually.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+
+.. code-block:: cpp
+
+   for (const auto& filename : dataset->files()) {
```
It seems a bit odd that here we iterate through and print `dataset->files()` when in an earlier example we were iterating `dataset->GetFragments()`.  Although I suppose I understand it is because we are working with the more specialized `FileSystemDataset` and not a plain old `Dataset`.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+Reading different file formats
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The above examples use Parquet files on local disk, but the Dataset API
```
I don't believe the `,` belongs here.

drin replied:
I think it is actually correct here (https://www.grammarly.com/blog/comma-before-but/)

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The above examples use Parquet files on local disk, but the Dataset API
+provides a consistent interface across multiple file formats and sources.
```
`multiple file formats and sources` -> `multiple file formats and filesystems`?  Not sure on this one.  Either way, you don't really expand on this point.  There are no examples of using a "different source" whatever that might be.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+Currently, Parquet, Feather / Arrow IPC, and CSV file formats are supported;
+more formats are planned in the future.
+
+If we save the table as a Feather file instead of Parquet files:
```
`as a Feather file` -> `as Feather files`

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+   :linenos:
+   :lineno-match:
+
+then we can read the Feather file using the same functions, but passing a
```
You're inconsistent about whether you start a continuing sentence with `...` or not.  I'm not sure which is more correct but you do it differently on line 93.

westonpace replied:
`Feather file using the same functions, but passing a` -> `Feather file by passing a`.

Also, I'm pretty sure this is the wrong snippet to be referencing here.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+:class:`arrow::dataset::FileFormat` objects have properties that control how
+they are read. For example::
```
`how they are read` -> `how files are read`.

westonpace replied:
Also, you use a `::` here but elsewhere you use a `:`.  But this may just be an `rst` thing I don't understand.

drin replied:
I also think that since the member variable is `reader_options`, I'd prefer to see:

`FileFormat objects have options that control read behavior`

...but only ParquetFileFormat has `reader_options`? Is this called something different in each `FileFormat` class?

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+Filtering data
+--------------
+
+So far, we've read the entire dataset, but this can be inefficient or use too
```
`we've read the entire dataset` -> `we've been reading the entire dataset`

Instead of `inefficient` can you be more specific?  Maybe `use too much memory or waste time reading data we don't need.`?

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+much memory. The :class:`arrow::dataset::Scanner` offers control over what data
+to read.
+
+:func:`arrow::dataset::ScannerBuilder::Project` lets us read only specified
```
Technically `Project` does more than just that but I think it's fine to omit that complexity at the moment.  Maybe change to `In this snippet we use Project to select which columns to read`.

drin replied:
I mostly agree with this, but my assumption is that `Project` selects which columns are produced? I don't know the implementation details, but I hope that something else actually determines which columns are read from the source (since a filter may be applied to a column not in the projection)?

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+   :linenos:
+   :lineno-match:
+
+For formats which support pushdown (such as Parquet), this means that only the
```
"pushdown" is maybe not a term known by the audience here.  Also, I don't think column selection is technically "pushdown".  Maybe `Some formats (such as Parquet) can reduce I/O costs by reading specific columns from the filesystem.`

drin replied:
maybe end with `...from the file.` since I think the filesystem is a slightly higher abstraction than the file? At least, as far as the implementations here. Generally speaking, the underlying file system defines whether parts of a file can be read into memory, but I don't think that's what's being referred to here.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+
+A filter can be provided with :func:`arrow::dataset::ScannerBuilder::Filter`,
+so that rows which do not match the filter predicate will not be included in
+the returned table.
```
Maybe add a sentence...

`  Some formats (such as Parquet) can use this filter to reduce the amount of I/O needed.`

drin replied:
> Maybe add a sentence...
> 
> ` Some formats (such as Parquet) can use this filter to reduce the amount of I/O needed.`

I think this type of note is only useful if there's a description of (or reference to) how the I/O is reduced. So, I think it would be better mentioned in format-specific documentation somewhere.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+
+In addition to selecting columns,
+:func:`arrow::dataset::ScannerBuilder::Project` can also be used for more
+complex projections in combination with expressions.
```
Drop `in combination with expressions.`

drin replied:
> Drop `in combination with expressions.`

Maybe just combine this sentence with the following one:

```
In addition to selecting columns, Project can take a vector of
expressions (for more complex projections) and a vector of
column names. This usage evaluates each expression to a
column value, associated with the co-indexed column name.
```

(newlines added for comment formatting)

I'm not sure the "co-indexing" is clear (or necessary) so feel free to remove/improve that part.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+:func:`arrow::dataset::ScannerBuilder::Project` can also be used for more
+complex projections in combination with expressions.
+
+In this case, we pass a vector of expressions used to construct column values,
```
Second comma on this line not needed.

westonpace commented on docs/source/cpp/dataset.rst:
The whole sentence starting with `However` is a little awkward to me.  Maybe...

```
However, a "partitioned" dataset relies on nested directory structures where sub-directory names hold describe the data stored in that directory
```

...but I'm not really sold on that either.

westonpace commented on docs/source/cpp/dataset.rst:
`may look like on disk` -> `may have the following layout`.

westonpace commented on docs/source/cpp/dataset.rst:
It feels a bit weird to have "Dataset" with a capital D and no backticks or link.  Could maybe phrase it as "Arrow's dataset writing functionality."

westonpace commented on docs/source/cpp/dataset.rst:
`uses` -> `should use`

westonpace commented on docs/source/cpp/dataset.rst:
`predicate` -> `filter`.

westonpace commented on docs/source/cpp/dataset.rst:
`the partition *factory*` -> `the hive partitioning factory`

westonpace commented on docs/source/cpp/dataset.rst:
"Directory partitioning" is kind of a confusing name when "partitioning" means "using directories".  Maybe to help with that we could rephrase this to...

```
Arrow supports another form of partitioning, "Directory partitioning", where the segments...
```

westonpace commented on docs/source/cpp/dataset.rst:
`name` -> `names` or `name are` -> `name is`.  I think both work.

westonpace commented on docs/source/cpp/dataset.rst:
Not sure but `interface` -> `API`?

westonpace commented on docs/source/cpp/dataset.rst:
`a` -> `an`

westonpace commented on docs/source/cpp/dataset.rst:
`to be used` -> `which was used`

westonpace commented on docs/source/cpp/dataset.rst:
`supports` -> `support`

westonpace commented on docs/source/python/dataset.rst:
```diff
@@ -38,9 +38,9 @@ tabular, potentially larger than memory and multi-file datasets:
-database connections).
+Currently, only Parquet, Feather / Arrow IPC, and CSV files are supported. The
+goal is to expand this in the future to other file formats and data sources
+(e.g.  database connections).
```
```
(e.g.  database
```
```
(e.g. database
```

westonpace reviewed (commented):
Hope you don't mind some nitpicking :)

lidavidm commented:
Thanks for the comments. I'll go through and make sure changes are consistent between Python/C++ (the C++ side of the docs was based on the Python one).

FWIW, the double-colon is an reST thing: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#literal-blocks

drin commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+    is not yet guaranteed.
+
+The Arrow Datasets library provides functionality to efficiently work with
+tabular, potentially larger than memory and multi-file datasets:
```
I don't think this is necessary

drin replied:
Actually, looking at the below section, maybe lift the first bullet point so that the datasets library isn't defined as a library for datasets? This suggestion includes @westonpace 's next few comments and the following bullet points:

```
The Arrow Datasets library provides a unified interface for operations on tabular data from various sources, in various formats. The data may be from buffers, local filesystems, and cloud filesystems; and may be larger than memory or span multiple units (e.g. multiple files). Some supported data formats include parquet, feather, and IPC. Some supported operations include discovery, predicate pushdowns, and read parallelism.
```

drin commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,381 @@
+* Optimized reading with predicate pushdown (filtering rows), projection
+  (selecting columns), parallel reading, or fine-grained managing of tasks.
+
+Currently, only Parquet, Feather / Arrow IPC, and CSV files are supported. The
```
Change `supported` -> `implemented`

drin replied:
also, if this is kept, then maybe the reference to data formats above should just be removed to reduce unnecessary repetition

lidavidm commented:
@westonpace (and/or @drin!) would you like to give a final read-through? I think this may cause minor conflicts with the rest of our PRs due to the new doc comments.

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,389 @@
+  Parquet, Feather / Arrow IPC, and CSV files) and different file systems (local,
+  cloud).
+* Discovery of sources (crawling directories, handling partitioned datasets with
+  various partitioning schemes, basic schema normalization, ..)
```
`..` -> `...`

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,389 @@
+In addition to searching a base directory, we can list file paths manually.
+
+Creating a :class:`arrow::dataset::Dataset` does not begin reading the data
+itself. If needed, it only crawls the directory to find all the files
```
`If needed, It only crawls...all the files` -> `It only crawls...all the files (if needed).`

westonpace commented on docs/source/cpp/dataset.rst:
```diff
@@ -0,0 +1,389 @@
+   :linenos:
+   :lineno-match:
+
+…then we can read the Feather file by passing a :class:`arrow::dataset::IpcFileFormat`:
```
`a` -> `an`

westonpace reviewed (commented):
LGTM

westonpace commented:
> I also think that since the member variable is reader_options, I'd prefer to see:
> 
> FileFormat objects have options that control read behavior
> 
> ...but only ParquetFileFormat has reader_options? Is this called something different in each FileFormat class?
> 
@drin

I'm not sure if it's called out here but the preferred way of setting options would probably be to supply them to the `ScannerBuilder` using the `FragmentScanOptions` method.  There are `IpcFragmentScanOptions`, `CsvFragmentScanOptions`, and `ParquetFragmentScanOptions`.

westonpace commented:
> I mostly agree with this, but my assumption is that Project selects which columns are produced? I don't know the implementation details, but I hope that something else actually determines which columns are read from the source (since a filter may be applied to a column not in the projection)?
@drin 

Yes, the columns read from the disk will be the based on the columns needed for the projection expression and the columns needed for the filter expression.

lidavidm commented:
I've updated https://lidavidm.github.io/arrow-docs-next/cpp/dataset.html#tabular-datasets. Thanks @westonpace for pointing out those comments I've missed. I've updated the text to address them.

lidavidm commented:
I'm going to merge, but @drin thanks for all your comments! If you have further feedback or run into other things, please feel free to follow up here or create a JIRA issue (https://issues.apache.org/jira/, feel free to tag @lidavidm there) and we can iterate further.

drin commented:
> I'm going to merge, but @drin thanks for all your comments! If you have further feedback or run into other things, please feel free to follow up here or create a JIRA issue (https://issues.apache.org/jira/, feel free to tag @lidavidm there) and we can iterate further.

sounds good, thanks and great job!
